### PR TITLE
Self variant enum res fix

### LINKED ISSUE #4789: `Self::Variant` patterns are not resolved correctly
```rust
enum Enum {
    Variant,
}

impl Enum {
    fn f(self) {
        match self {
            Enum::Variant => {}
        }
    }

    fn g(self) {
        match self {
            Self::Variant => {}
        }
    }
}
```

In `f`, everything works correctly. In `g`, `Variant` is not recognized correctly and does not show up in completions after `Self::`.

(I'm kind of assuming that this is a known bug, but I couldn't find an issue about it, so I'm opening this to track it)

### PR DESCRIPTION
**Pull request by BGluth**
Fixes #4789.

This is my first PR for this project, so it's probably worth giving it an extra close look.

A few things that I wasn't sure about:
- Is `resolve_path` really the best place to perform this check? It seemed like a natural place, but perhaps there's a better place?
- When handling the new variant `PathResolution::VariantDef`, I couldn't see an obvious variant of `TypeNs` to return in `in_type_ns` for Unions and Structs.

### CODE REVIEW

File: crates/ra_ide_db/src/defs.rs [hunk 1]
```diff
@@ -276,6 +276,14 @@ pub fn classify_name_ref(
         PathResolution::TypeParam(par) => Definition::TypeParam(par),
         PathResolution::Macro(def) => Definition::Macro(def),
         PathResolution::SelfType(impl_def) => Definition::SelfType(impl_def),
+        PathResolution::VariantDef(variant) => {
+            let def = match variant {
```
Comment (flodiebold) [inline crates/ra_ide_db/src/defs.rs:280]: we should just add an `impl From<VariantDef> for ModuleDef`.

File: crates/ra_hir/src/semantics.rs [hunk 2]
```diff
@@ -40,6 +40,7 @@ pub enum PathResolution {
     SelfType(ImplDef),
     Macro(MacroDef),
     AssocItem(AssocItem),
+    VariantDef(VariantDef),
```
[the remark this answers is not in the document]
Comment (BGluth) [reply crates/ra_hir/src/semantics.rs:43]: Agreed.

**Comment by BGluth** [comment]
I'll see if I can also handle the record case in this PR. @flodiebold 

Just as a sanity check, for the record case, you mean something like this should resolve right?
```rust
enum Foo {
    Bar{baz: usize},
}

impl Foo {
    fn baz(self) {
        match self {
            Self::Bar { baz<|> } => {}
        }
    }
}
```

It seems that even without this PR `baz` resolves fine. With the PR `Self::Bar` also resolves.

**Comment by BGluth** [comment]
The level of code duplication here that I introduced is bothering me a bit:
```rust
if let Some(path_expr) = path.syntax().parent().and_then(ast::PathExpr::cast) {
    let expr_id = self.expr_id(db, &path_expr.into())?;
    if let Some(assoc) = self.infer.as_ref()?.assoc_resolutions_for_expr(expr_id) {
        return Some(PathResolution::AssocItem(assoc.into()));
    }
    if let Some(VariantId::EnumVariantId(variant)) =
        self.infer.as_ref()?.variant_resolution_for_expr(expr_id)
    {
        return Some(PathResolution::Def(ModuleDef::EnumVariant(variant.into())));
    }
}
if let Some(path_pat) = path.syntax().parent().and_then(ast::PathPat::cast) {
    let pat_id = self.pat_id(&path_pat.into())?;
    if let Some(assoc) = self.infer.as_ref()?.assoc_resolutions_for_pat(pat_id) {
        return Some(PathResolution::AssocItem(assoc.into()));
    }
    if let Some(VariantId::EnumVariantId(variant)) =
        self.infer.as_ref()?.variant_resolution_for_pat(pat_id)
    {
        return Some(PathResolution::Def(ModuleDef::EnumVariant(variant.into())));
    }
}
```

I would normally put the shared stuff into a function something like this:
```rust
fn resolve_with_infer<T>(&self, id: Idx<T>) -> Option<PathResolution> {
...
}
```

But `Idx` is defined in `ra_arena` which is not a dependency of the `ra_hir` crate, and I'm not sure if it's a good idea to add a dependency just for this.

File: crates/ra_ide_db/src/defs.rs (hunk 1 above)
Comment (BGluth) [reply crates/ra_ide_db/src/defs.rs]: Yeah I agree this is cleaner in this case.

**Comment by flodiebold** [comment]
I think that's fine -- it's good to directly see all the different cases being handled. Can you add tests for the record cases? E.g. what you wrote for patterns, and something like
```rust
impl Foo {
    fn baz(self) {
        Self::Bar<|> {}
    }
}
```
for expressions.

**Comment by BGluth** [comment]
Yeah not a problem!

I'm just out of town for another day but I'll add the tests as soon as I get back.

**Comment by BGluth** [comment]
Just a little update.

I have four tests for the pattern/expression and enum const/record cases, and I've realized it's not resolving the record cases correctly yet. For example:

```rust
enum Foo {
    Bar { val: i32 },
}
impl Foo {
    fn baz(self) {
        Self::Bar<|> {val: 4};
    }
}
```
This doesn't resolve (`Self::Bar` is a record literal). What's kind of interesting is if we are explicit with the enum variant (replace `Self` with `Foo`) it ends up resolving in `resolve_hir_path`, even though it appears in `record_field_resolutions` from type inference. So I guess it's probably just a matter of figuring out where to check `record_field_resolutions` and `record_field_pat_resolutions` (should the explicit emum variant case even be resolving in `resolve_hir_path`?).

**Comment by flodiebold** [comment]
Well, neither should be resolving from `record_field_resolutions` or `record_field_pat_resolutions`, those are for fields. They should be resolving from `variant_resolutions`. Maybe it's not checking the right expression?

**Comment by flodiebold** [comment]
The code only looks in `variant_resolutions` for an `ast::PathExpr`, but in the record case there is no `PathExpr`, but an `ast::RecordLit`. Similar for patterns.

**Comment by BGluth** [comment]
Yeah looking closer, it looks like there's only one segment of whatever is resolving in `record_field_pat_resolutions`, which I guess probably just the field `val` and not `Foo::Bar`. I should have mentioned that it's getting resolved in `variant_resolutions`.
```
record_field_pat_resolutions: {
    Idx::<Pat>(1): FieldId {
        parent: EnumVariantId(
            EnumVariantId {
                parent: EnumId(
                    0,
                ),
                local_id: Idx::<EnumVariantData>(0),
            },
        ),
        local_id: Idx::<FieldData>(0),
    },
},
variant_resolutions: {
    PatId(
        Idx::<Pat>(2),
    ): EnumVariantId(
        EnumVariantId {
            parent: EnumId(
                0,
            ),
            local_id: Idx::<EnumVariantData>(0),
        },
    ),
},
```

Any thoughts on a good location to perform these checks? Should I create a new function somewhat similar to `resolve_path` but just deals with records?

**Comment by BGluth** [comment]
The more I think about it doing the checks in `resolve_path` seems like a good choice.

**Comment by BGluth** [comment]
If `resolve_path` is a good spot, I've made the changes and all tests are now passing.

File: crates/ra_hir/src/source_analyzer.rs [hunk 3]
```diff
@@ -216,13 +216,49 @@ impl SourceAnalyzer {
...
+            {
+                return Some(PathResolution::Def(ModuleDef::EnumVariant(variant.into())));
+            }
+        }
+
+        if let Some(rec_lit) = path.syntax().parent().and_then(ast::RecordLit::cast) {
+            let expr_id = self.expr_id(db, &rec_lit.into())?;
+            if let Some(assoc) = self.infer.as_ref()?.assoc_resolutions_for_expr(expr_id) {
+                return Some(PathResolution::AssocItem(assoc.into()));
+            }
```
Comment (flodiebold) [inline crates/ra_hir/src/source_analyzer.rs:242]: I don't think we need these three lines

File: crates/ra_hir/src/source_analyzer.rs [hunk 4]
```diff
@@ -216,13 +216,49 @@ impl SourceAnalyzer {
...
+            {
+                return Some(PathResolution::Def(ModuleDef::EnumVariant(variant.into())));
+            }
+        }
+
+        if let Some(rec_pat) = path.syntax().parent().and_then(ast::RecordPat::cast) {
+            let pat_id = self.pat_id(&rec_pat.into())?;
+            if let Some(assoc) = self.infer.as_ref()?.assoc_resolutions_for_pat(pat_id) {
+                return Some(PathResolution::AssocItem(assoc.into()));
+            }
```
Comment (flodiebold) [inline crates/ra_hir/src/source_analyzer.rs:254]: ... or these

**Comment by flodiebold** [comment]
LGTM except for some unnecessary cases!

bors delegate+

File: crates/ra_hir/src/source_analyzer.rs (hunk 3 above)
Comment (BGluth) [reply crates/ra_hir/src/source_analyzer.rs]: Removed.

File: crates/ra_hir/src/source_analyzer.rs (hunk 4 above)
Comment (BGluth) [reply crates/ra_hir/src/source_analyzer.rs]: Yep.

**Pull request merged**